Cython/Utility/CommonStructures.c

```diff
@@ -66,19 +75,30 @@ static PyTypeObject *__Pyx_FetchCommonTypeFromSpec(PyObject *module, PyType_Spec
             goto bad;
         }
         goto done;
+    } else if (unlikely(get_item_ref_result == -1)) {
+        goto bad;
     }
 
-    if (!PyErr_ExceptionMatches(PyExc_AttributeError)) goto bad;
-    PyErr_Clear();
     // We pass the ABI module reference to avoid keeping the user module alive by foreign type usages.
     CYTHON_UNUSED_VAR(module);
     cached_type = __Pyx_PyType_FromModuleAndSpec(abi_module, spec, bases);
     if (unlikely(!cached_type)) goto bad;
     if (unlikely(__Pyx_fix_up_extension_type_from_spec(spec, (PyTypeObject *) cached_type) < 0)) goto bad;
-    if (PyObject_SetAttrString(abi_module, object_name, cached_type) < 0) goto bad;
+
+    new_cached_type = __Pyx_PyDict_SetDefault(abi_module_dict, py_object_name, cached_type, 1);
+    if (unlikely(!new_cached_type)) goto bad;
+    if (unlikely(new_cached_type != cached_type)) {
+        // race to initialize it - use the value that's already been set.
+        Py_DECREF(cached_type);
+        cached_type = new_cached_type;
+        goto verify_type;
```

Comment:
Could `verify_type` be a separate function, called twice if necessary?

Reply:
It already kind of is - there was just a chunk of code left out of it for some reason. I've hopefully improved that.